### PR TITLE
feat:(picker): add aria describedby action menu, picker, split button

### PR DESCRIPTION
**Pull request by blunteshwar**
## Description
 Apply aria-describedby content to Action Button, Picker, Split Button]

## Related issue(s)

-

## How has this been tested?
Using accessibility and inspect element

-   [ ] _Test case 1_
    1. Go here
    2. Do this
-   [ ] _Test case 2_
    1. Go here
    2. Do this

## Types of changes

-   [ ] Bug fix (non-breaking change which fixes an issue)
-   [x] New feature (non-breaking change which adds functionality)
-   [ ] Breaking change (fix or feature that would cause existing functionality to change)
-   [ ] Chore (minor updates related to the tooling or maintenance of the repository, does not impact compiled assets)

## Checklist

-   [x] I have signed the [Adobe Open Source CLA](http://opensource.adobe.com/cla.html).
-   [x] My code follows the code style of this project.
-   [x] If my change required a change to the documentation, I have updated the documentation in this pull request.
-   [x] I have read the **[CONTRIBUTING](<(https://github.com/adobe/spectrum-web-components/blob/main/CONTRIBUTING.md)>)** document.
-   [x] I have added tests to cover my changes.
-   [x] All new and existing tests passed.
-   [x] I have reviewed at the Accessibility Practices for this feature, see: [Aria Practices](https://www.w3.org/TR/wai-aria-practices/)

## Best practices

This repository uses conventional commit syntax for each commit message; note that the GitHub UI does not use this by default so be cautious when accepting suggested changes. Avoid the "Update branch" button on the pull request and opt instead for rebasing your branch against `main`.

### CODE REVIEW

File: packages/split-button/stories/index.ts [hunk 1]
```diff
@@ -156,6 +157,7 @@ export const splitbutton = (
         type=${properties.type || 'field'}
         ?open=${!!properties.open}
         ?disabled=${properties.disabled}
+        description=${properties.description}
```
Comment (Rajdeepc) [inline packages/split-button/stories/index.ts:160]: please use a ifDefined here too

**Review by Rajdeepc — CHANGES_REQUESTED** [review]

File: packages/split-button/stories/index.ts (hunk 1 above)
Comment (blunteshwar) [reply packages/split-button/stories/index.ts:160]: here description is an inherited property, it's inherited from `picker` hence I believe we don't explicitly need `ifDefined`

**Review by Rajdeepc — APPROVED** [review]

File: packages/action-button/stories/index.ts [hunk 2]
```diff
@@ -42,6 +43,7 @@ export function renderButton(properties: Properties): TemplateResult {
             ?disabled=${!!properties.disabled}
             ?selected=${!!properties.selected}
             ?toggles=${!!properties.toggles}
+            description="${ifDefined(properties.description)}"
```
[the remark this answers is not in the document]
Comment (blunteshwar) [reply packages/action-button/stories/index.ts]: I have tried to correct this according to my understanding
Please have a look

**Comment by Westbrook** [comment]
Looks like this branch is pretty far behind `main`. Please catch it up and ensure that all VRTs are passing before we dive into code review.

**Comment by blunteshwar** [comment]
@Westbrook the code coverage on my local is passing but here its not passing.

**Comment by Westbrook** [comment]
It's not "failing", it's telling you that you added more code than you tested and falling below coverage levels: https://app.circleci.com/pipelines/github/adobe/spectrum-web-components/17756/workflows/04b029be-f42c-4014-8ca2-c8a7363c8a86/jobs/404448

File: packages/picker/src/Picker.ts [hunk 3]
```diff
@@ -60,6 +60,18 @@ const chevronClass = {
...
+    static instanceCount = 0;
+    instanceCount: number;
+
+    @property({ type: String })
+    public description = '';
+
+    constructor() {
+        super();
+        this.instanceCount = PickerBase.instanceCount++;
+        this.description = `sp-picker-${this.instanceCount}`;
```
Comment (Westbrook) [inline packages/picker/src/Picker.ts:72]: This description is being completed within the shadow root, we should be able to use a static value here, rather than constructing something with the instance count.

File: packages/color-area/test/color-area.test.ts [hunk 4]
```diff
@@ -547,6 +547,8 @@ describe('ColorArea', () => {
         );
 
         await elementUpdated(el);
+        const Xvalue = Number(Number(el.inputX.value).toFixed(2));
+        const Yvalue = Number(Number(el.inputY.value).toFixed(2));
```
Comment (Westbrook) [inline packages/color-area/test/color-area.test.ts:551]: Can we fast forward unrelated changes into a standalone PR that we can merge first?

File: packages/split-button/src/SplitButton.ts [hunk 5]
```diff
@@ -139,6 +140,7 @@ export class SplitButton extends SizedMixin(PickerBase) {
             `,
             html`
                 <sp-button
+                    aria-describedby=${this.description}
```
Comment (Westbrook) [inline packages/split-button/src/SplitButton.ts:143]: This ones hard to make decisions about, as it's deprecated. Hard to say whether we should be describing both buttons in this way or not.

File: packages/action-menu/src/ActionMenu.ts [hunk 6]
```diff
@@ -87,6 +87,7 @@ export class ActionMenu extends ObserveSlotPresence(
     protected override render(): TemplateResult {
         return html`
             <sp-action-button
+                aria-describedby=${this.description}
```
Comment (Westbrook) [inline packages/action-menu/src/ActionMenu.ts:90]: You don't seem to be making this same associating in the `Picker.ts` code...

File: packages/action-button/stories/index.ts [hunk 7]
```diff
@@ -42,6 +42,9 @@ export function renderButton(properties: Properties): TemplateResult {
             ?disabled=${!!properties.disabled}
             ?selected=${!!properties.selected}
             ?toggles=${!!properties.toggles}
+
+            description="${properties.description}"
+
```
Comment (Westbrook) [inline packages/action-button/stories/index.ts:47]: Based on current usage, e.g. the component builds its own value for this property, do we want to surface it in the stories?

File: packages/color-area/test/color-area.test.ts (hunk 4 above)
Comment (blunteshwar) [reply packages/color-area/test/color-area.test.ts:551]: These are just some tests for color-area that I added to increase the code coverage. Other than this I didn't make any changes in the color-area component!

Comment (Rajdeepc) [reply packages/color-area/test/color-area.test.ts:551]: @blunteshwar I think Westbrook is right! lets not pollute this PR with color-area changes.

File: packages/action-button/stories/index.ts (hunk 7 above)
Comment (blunteshwar) [reply packages/action-button/stories/index.ts:47]: removed from stories
Done!

File: packages/picker/test/index.ts [hunk 8]
```diff
@@ -248,6 +248,11 @@ export function runPickerTests(): void {
         it('loads accessibly', async () => {
             await expect(el).to.be.accessible();
         });
+        it('checks for elements id and  description', async () => {
+            el = await pickerFixture();
+            expect(el).to.have.attribute('id');
+            expect(el.description).to.not.be.null;
```
[the remark this answers is not in the document]
Comment (blunteshwar) [reply packages/picker/test/index.ts:254]: done!

File: packages/action-menu/src/ActionMenu.ts (hunk 6 above)
Comment (blunteshwar) [reply packages/action-menu/src/ActionMenu.ts:90]: done!

File: packages/picker/src/Picker.ts (hunk 3 above)
Comment (blunteshwar) [reply packages/picker/src/Picker.ts]: done!

File: packages/split-button/src/SplitButton.ts (hunk 5 above)
Comment (blunteshwar) [reply packages/split-button/src/SplitButton.ts:143]: @Rajdeepc can you help me out with this?

File: packages/picker/src/Picker.ts [hunk 9]
```diff
@@ -417,6 +425,13 @@ export class PickerBase extends SizedMixin(Focusable, { noDefaultSize: true }) {
         `;
     }
 
+    public renderDescriptionSlot(): TemplateResult {
+        return html`
+            <div id=${ifDefined(this.description)}>
```
Comment (Westbrook) [inline packages/picker/src/Picker.ts:430]: This value is internal, please make it static. Doing so should allow us to remove the `description` property from this class as well.

File: packages/picker/src/Picker.ts [hunk 10]
```diff
@@ -417,6 +425,13 @@ export class PickerBase extends SizedMixin(Focusable, { noDefaultSize: true }) {
         `;
     }
 
+    public renderDescriptionSlot(): TemplateResult {
```
Comment (Westbrook) [inline packages/picker/src/Picker.ts:428]: Please make this a getter to match with other sub-templates in this file. It should also be `private` (if only used in this file) or `protected` (if leveraged in other class extensions).

File: packages/action-menu/test/index.ts [hunk 11]
```diff
@@ -304,6 +304,15 @@ export const testActionMenu = (mode: 'sync' | 'async'): void => {
             expect(button).to.have.attribute('aria-expanded', 'true');
             expect(button).to.have.attribute('aria-controls', 'menu');
         });
+        it('has attribute aria-describedby', async () => {
```
Comment (Westbrook) [inline packages/action-menu/test/index.ts:307]: Great first step, but what we really need here is a test leveraged [findDescribedNode](https://github.com/adobe/spectrum-web-components/blob/main/test/testing-helpers-a11y.ts#L21) to confirm the expected association is being created here.

File: packages/picker/src/Picker.ts (hunk 9 above)
Comment (Westbrook) [reply packages/picker/src/Picker.ts:430]: Once this value is static, please add documentation for this slot similar to [this](https://github.com/adobe/spectrum-web-components/blob/main/packages/picker/src/Picker.ts#L689).

File: packages/picker/src/Picker.ts (hunk 10 above)
Comment (blunteshwar) [reply packages/picker/src/Picker.ts:428]: Done!

File: packages/picker/src/Picker.ts (hunk 9 above)
Comment (blunteshwar) [reply packages/picker/src/Picker.ts:430]: Done!

File: packages/action-menu/test/index.ts (hunk 11 above)
Comment (blunteshwar) [reply packages/action-menu/test/index.ts:308]: Done!

File: packages/picker/src/Picker.ts [hunk 12]
```diff
@@ -60,6 +60,13 @@ const chevronClass = {
 export class PickerBase extends SizedMixin(Focusable, { noDefaultSize: true }) {
     protected isMobile = new MatchMediaController(this, IS_MOBILE);
 
+    @property({ type: String })
+    public description = 'sp-picker';
+
+    constructor() {
+        super();
+    }
```
Comment (Westbrook) [inline packages/picker/src/Picker.ts:68]: Please no empty methods.

File: packages/picker/src/Picker.ts [hunk 13]
```diff
@@ -417,6 +424,13 @@ export class PickerBase extends SizedMixin(Focusable, { noDefaultSize: true }) {
         `;
     }
 
+    protected get renderDescriptionSlot(): TemplateResult {
+        return html`
+            <div id="sp-picker">
```
Comment (Westbrook) [inline packages/picker/src/Picker.ts:429]: I think we should use `description` or similar here, so that when it is leveraged in Action Menu or Split Button, no one is surprised.

File: packages/picker/src/Picker.ts [hunk 14]
```diff
@@ -60,6 +60,13 @@ const chevronClass = {
 export class PickerBase extends SizedMixin(Focusable, { noDefaultSize: true }) {
     protected isMobile = new MatchMediaController(this, IS_MOBILE);
 
+    @property({ type: String })
+    public description = 'sp-picker';
```
Comment (Westbrook) [inline packages/picker/src/Picker.ts:64]: I'm not sure we need to cary this as a property being we're leveraging a static value in 429, below.

File: packages/action-menu/src/ActionMenu.ts [hunk 15]
```diff
@@ -87,6 +87,7 @@ export class ActionMenu extends ObserveSlotPresence(
     protected override render(): TemplateResult {
         return html`
             <sp-action-button
+                aria-describedby=${ifDefined(this.description)}
```
Comment (Westbrook) [inline packages/action-menu/src/ActionMenu.ts:90]: This should be statically applied, `ifDefined()` shouldn't be needed.

**Review by Westbrook — COMMENTED** [review]
Please associate this content in Picker and Split Button, too.

File: packages/picker/src/Picker.ts (hunk 12 above)
Comment (blunteshwar) [reply packages/picker/src/Picker.ts:68]: Done!

File: packages/picker/src/Picker.ts (hunk 13 above)
Comment (blunteshwar) [reply packages/picker/src/Picker.ts:429]: Done!

File: packages/picker/src/Picker.ts (hunk 14 above)
Comment (blunteshwar) [reply packages/picker/src/Picker.ts:64]: I used it as a property because action-menu is leveraging this by inheriting it from picker.
If we declare description merely as a static value then I believe action-menu won't be able to inherit it and leverage it

File: packages/action-menu/src/ActionMenu.ts (hunk 15 above)
Comment (blunteshwar) [reply packages/action-menu/src/ActionMenu.ts:90]: Done!

File: packages/picker/src/Picker.ts (hunk 14 above)
Comment (Westbrook) [reply packages/picker/src/Picker.ts:64]: Could you expand on why you want this value settable from the outside?

Comment (blunteshwar) [reply packages/picker/src/Picker.ts:64]: I don't want this value to be settable from outside.
Shall I do like this

```
protected _description= 'sp-picker';

  static get description(): string {
    return this._description;
  }
```

Comment (Westbrook) [reply packages/picker/src/Picker.ts:64]: This works, you can also hand it directly on the class for a little less code.
```
protected description = 'sp-picker';
```

Historically, the library has simply leveraged these values as magic strings, which would be acceptable here. This means we "know" via convention that the "description" of an element will be id'd as `description` rather than marking that out in sharable code.

However, we do have [this issue](https://github.com/adobe/spectrum-web-components/issues/2064) open to revisit that.

As a "constant", it may be of use to create and export one from this file, so it's not fixedly related to the class:
```
export const DESCRIPTION_ID = 'description';
```

Related, I wanted to work out how we hold that value before questioning the value itself, but the element is _sometimes_ called `sp-picker`, I'm not sure the description should be id'd as such...something a little more clear to what it does would be nice.
